Add enrichment + check suppression support for plan file scans

ac-square opened the pull request:
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

This PR introduces the `--repo_root_for_plan_enrichment` flag which, when used in conjunction with `-f` on a plan file, will enrich the reports with the appropriate filepath, line numbers, and codeblock of the hcl code that generated the plan file.

This flag should be passed a repo root of wherever the hcl files live.

When not set (which it isn't by default), the functionality for scanning a plan file continues as normal.

``` checkov -f ~/Development/tf-test/planfile.json --repo-root-for-plan-enrichment ~/Development/tf-test```

ac-square commented:
[tf-test.zip](https://github.com/bridgecrewio/checkov/files/6498297/tf-test.zip)

Running `checkov -f ./tf-test/planfile.json --repo-root-for-plan-enrichment ./tf-test`

```hcl
terraform_plan scan results:

Passed checks: 4, Failed checks: 8, Skipped checks: 0

Check: CKV_AWS_62: "Ensure IAM policies that allow full "*-*" administrative privileges are not created"
	PASSED for resource: aws_iam_policy.policy
	File: /tf-test/planfile.json:0-0
	Guide: https://docs.bridgecrew.io/docs/iam_47

Check: CKV_AWS_18: "Ensure the S3 bucket has access logging enabled"
	PASSED for resource: aws_s3_bucket.b
	File: /tf-test/planfile.json:0-0
	Guide: https://docs.bridgecrew.io/docs/s3_13-enable-logging

Check: CKV_AWS_70: "Ensure S3 bucket does not allow an action with any Principal"
	PASSED for resource: aws_s3_bucket.b
	File: /tf-test/planfile.json:0-0
	Guide: https://docs.bridgecrew.io/docs/bc_aws_s3_23

Check: CKV_AWS_93: "Ensure S3 bucket policy does not lockout all but root user. (Prevent lockouts needing root account fixes)"
	PASSED for resource: aws_s3_bucket.b
	File: /tf-test/planfile.json:0-0
	Guide: https://docs.bridgecrew.io/docs/bc_aws_iam_53

Check: CKV_AWS_63: "Ensure no IAM policies documents allow "*" as a statement's actions"
	FAILED for resource: aws_iam_policy.policy
	File: main.tf:1-19
	Guide: https://docs.bridgecrew.io/docs/iam_48

		1  | resource "aws_iam_policy" "policy" {
		2  |   name        = "my_policy-${var.aws_account_id}"
		3  |   path        = "/"
		4  |   description = "My test policy"
		5  |   policy = <<EOF
		6  | {
		7  |   "Version": "2012-10-17",
		8  |   "Statement": [
		9  |     {
		10 |       "Action": [
		11 |         "*"
		12 |       ],
		13 |       "Effect": "Allow",
		14 |       "Resource": "arn:aws:iam::${var.aws_account_id}:role/admin"
		15 |     }
		16 |   ]
		17 | }
		18 | EOF
		19 | }


Check: CKV_AWS_19: "Ensure all data stored in the S3 bucket is securely encrypted at rest"
	FAILED for resource: aws_s3_bucket.b
	File: main.tf:21-30
	Guide: https://docs.bridgecrew.io/docs/s3_14-data-encrypted-at-rest

		21 | resource "aws_s3_bucket" "b" {
		22 |   bucket = "my-tf-test-bucket"
		23 |   # checkov:skip=CKV_AWS_57: The bucket is a public static content host
		24 |   acl    = "public-read-write"
		25 |
		26 |   tags = {
		27 |     Name        = "My bucket"
		28 |     Environment = "Dev"
		29 |   }
		30 | }


Check: CKV_AWS_20: "S3 Bucket has an ACL defined which allows public READ access."
	FAILED for resource: aws_s3_bucket.b
	File: main.tf:21-30
	Guide: https://docs.bridgecrew.io/docs/s3_1-acl-read-permissions-everyone

		21 | resource "aws_s3_bucket" "b" {
		22 |   bucket = "my-tf-test-bucket"
		23 |   # checkov:skip=CKV_AWS_57: The bucket is a public static content host
		24 |   acl    = "public-read-write"
		25 |
		26 |   tags = {
		27 |     Name        = "My bucket"
		28 |     Environment = "Dev"
		29 |   }
		30 | }


Check: CKV_AWS_144: "Ensure that S3 bucket has cross-region replication enabled"
	FAILED for resource: aws_s3_bucket.b
	File: main.tf:21-30
	Guide: https://docs.bridgecrew.io/docs/ensure-that-s3-bucket-has-cross-region-replication-enabled

		21 | resource "aws_s3_bucket" "b" {
		22 |   bucket = "my-tf-test-bucket"
		23 |   # checkov:skip=CKV_AWS_57: The bucket is a public static content host
		24 |   acl    = "public-read-write"
		25 |
		26 |   tags = {
		27 |     Name        = "My bucket"
		28 |     Environment = "Dev"
		29 |   }
		30 | }


Check: CKV_AWS_145: "Ensure that S3 buckets are encrypted with KMS by default"
	FAILED for resource: aws_s3_bucket.b
	File: main.tf:21-30
	Guide: https://docs.bridgecrew.io/docs/ensure-that-s3-buckets-are-encrypted-with-kms-by-default

		21 | resource "aws_s3_bucket" "b" {
		22 |   bucket = "my-tf-test-bucket"
		23 |   # checkov:skip=CKV_AWS_57: The bucket is a public static content host
		24 |   acl    = "public-read-write"
		25 |
		26 |   tags = {
		27 |     Name        = "My bucket"
		28 |     Environment = "Dev"
		29 |   }
		30 | }


Check: CKV_AWS_57: "S3 Bucket has an ACL defined which allows public WRITE access."
	FAILED for resource: aws_s3_bucket.b
	File: main.tf:21-30
	Guide: https://docs.bridgecrew.io/docs/s3_2-acl-write-permissions-everyone

		21 | resource "aws_s3_bucket" "b" {
		22 |   bucket = "my-tf-test-bucket"
		23 |   # checkov:skip=CKV_AWS_57: The bucket is a public static content host
		24 |   acl    = "public-read-write"
		25 |
		26 |   tags = {
		27 |     Name        = "My bucket"
		28 |     Environment = "Dev"
		29 |   }
		30 | }


Check: CKV_AWS_52: "Ensure S3 bucket has MFA delete enabled"
	FAILED for resource: aws_s3_bucket.b
	File: main.tf:21-30

		21 | resource "aws_s3_bucket" "b" {
		22 |   bucket = "my-tf-test-bucket"
		23 |   # checkov:skip=CKV_AWS_57: The bucket is a public static content host
		24 |   acl    = "public-read-write"
		25 |
		26 |   tags = {
		27 |     Name        = "My bucket"
		28 |     Environment = "Dev"
		29 |   }
		30 | }


Check: CKV_AWS_21: "Ensure all data stored in the S3 bucket have versioning enabled"
	FAILED for resource: aws_s3_bucket.b
	File: main.tf:21-30
	Guide: https://docs.bridgecrew.io/docs/s3_16-enable-versioning

		21 | resource "aws_s3_bucket" "b" {
		22 |   bucket = "my-tf-test-bucket"
		23 |   # checkov:skip=CKV_AWS_57: The bucket is a public static content host
		24 |   acl    = "public-read-write"
		25 |
		26 |   tags = {
		27 |     Name        = "My bucket"
		28 |     Environment = "Dev"
		29 |   }
		30 | }
```

christophetd commented:
This looks interesting! Would this PR allow to suppress checks for specific resources using inline suppression? https://www.checkov.io/2.Basics/Suppressing%20and%20Skipping%20Policies.html So far, it is only supported when scanning HCL code (not the plan), c.f. #1224

schosterbarak commented:
this is very interesting. would that also work with external modules?

ac-square commented:
> This looks interesting! Would this PR allow to suppress checks for specific resources using inline suppression? https://www.checkov.io/2.Basics/Suppressing%20and%20Skipping%20Policies.html So far, it is only supported when scanning HCL code (not the plan), c.f. #1224

I have that functionality ready to go. :) Just haven't pushed it yet.

ac-square commented:
> this is very interesting. would that also work with external modules?

Not unless you somehow point this to the repo_root for the external module.

ac-square commented:
@schosterbarak I do wonder if it would be useful to call out in the results if a resource is from  an external module or not. For Square and how we integrate Checkov into CI, we currently ignore external modules when presenting checkov results to end users because the modules are usually owned by another team (or are third party modules).

schosterbarak commented:
> ignore external modules when presenting checkov results to end users because the modules are usually owned by another team (or are third

@ac-square what do you show for plan scanning results if a resource is sourced from an external module? how do you point back to the code today? or is it just dropped from the results?

schosterbarak commented:
@ac-square any chance you have unit tests on this one? if not I can try to help creating some.

ac-square commented:
> > ignore external modules when presenting checkov results to end users because the modules are usually owned by another team (or are third
> 
> @ac-square what do you show for plan scanning results if a resource is sourced from an external module? how do you point back to the code today? or is it just dropped from the results?

At Square we just drop them, but for this PR, they will be treated as normal results, just without the filepath/codeblock enrichment.

ac-square commented:
> @ac-square any chance you have unit tests on this one? if not I can try to help creating some.

I'll look into this today.

schosterbarak commented:
> > > ignore external modules when presenting checkov results to end users because the modules are usually owned by another team (or are third
> > 
> > 
> > @ac-square what do you show for plan scanning results if a resource is sourced from an external module? how do you point back to the code today? or is it just dropped from the results?
> 
> At Square we just drop them, but for this PR, they will be treated as normal results, just without the filepath/codeblock enrichment.

sounds good. thanks!

gruebel commented on checkov/common/runners/runner_registry.py:
```diff
@@ -99,3 +114,74 @@ def enrich_report_with_guidelines(scan_report, guidelines):
+    def enrich_plan_report(report, enriched_resources):
+        # This enriches reports with the appropriate filepath, line numbers, and codeblock
+        for record in report.failed_checks:
+            if record.resource in enriched_resources:
+                record.file_path = enriched_resources[record.resource]["scanned_file"]
+                record.file_line_range = enriched_resources[record.resource][
+                    "entity_lines_range"
+                ]
+                record.code_block = enriched_resources[record.resource][
+                    "entity_code_lines"
+                ]
```
instead of retrieving several times the same value from the dict you can change it to retrieve it once
```python
        for record in report.failed_checks:
            enriched_resource = enriched_resources.get(record.resource)
            if enriched_resource:
                record.file_path = enriched_resource["scanned_file"]
                record.file_line_range = enriched_resource["entity_lines_range"]
                record.code_block = enriched_resource[entity_code_lines"]
```
when this would be a Python 3.8+ project, then you could even use the walrus operator `:=` 😄

schosterbarak replied:
:) will take time to support python 3.8 only. 
We saw some envs with python 3.6 or lower that needed checkov.

gruebel replied:
let's go directly up to 3.10 only 😄 
@schosterbarak I hope it is ok, I did the code review here?

schosterbarak replied:
always good to have another set of eyes. Maybe we can migrate some of the comments to a github suggestion: http://haacked.com/archive/2019/06/03/suggested-changes/ ? 
It might make the changes slightly easier.

gruebel replied:
you are right, some of them can be made as a suggestion, didn't think about using suggestions 😄

gruebel commented on checkov/common/runners/runner_registry.py:
```diff
@@ -1,10 +1,19 @@
 import logging
+import os
 from abc import abstractmethod
+from collections import defaultdict
```
```suggestion
```

gruebel commented on checkov/common/runners/runner_registry.py:
```diff
@@ -99,3 +114,74 @@ def enrich_report_with_guidelines(scan_report, guidelines):
+        )
+
+        enriched_resources = defaultdict(dict)
+        for definition in tf_definitions.items():
+            definitions_context = parser_registry.enrich_definitions_context(definition)
+
+        for full_file_path, definition in tf_definitions.items():
+            abs_scanned_file, _ = tf_runner._strip_module_referrer(full_file_path)
```
Can be combined to

```suggestion
        for full_file_path, definition in tf_definitions.items():
            definitions_context = parser_registry.enrich_definitions_context((full_file_path, definition))
            abs_scanned_file, _ = tf_runner._strip_module_referrer(full_file_path)
```

gruebel commented on checkov/common/runners/runner_registry.py:
```diff
@@ -99,3 +114,74 @@ def enrich_report_with_guidelines(scan_report, guidelines):
+
+    @staticmethod
+    def get_enriched_resources(report, repo_root):
+        parser = Parser()
+        tf_definitions = {}
+        parsing_errors = {}
+        Parser.parse_directory(
+            parser,
+            directory=repo_root,  # assume plan file is in the repo-root
+            out_definitions=tf_definitions,
+            out_parsing_errors=parsing_errors,
+        )
```
you initalize the `Parser` class here and then pass it as a parameter into a method of the class which is not annotated as `@classmethod` please don't do this. Just use directly `parser.parse_directory(directory=...)
```suggestion
        tf_definitions = {}
        parsing_errors = {}
        Parser().parse_directory(
            directory=repo_root,  # assume plan file is in the repo-root
            out_definitions=tf_definitions,
            out_parsing_errors=parsing_errors,
        )
```

ac-square commented on checkov/common/runners/runner_registry.py:
```diff
@@ -99,3 +114,74 @@ def enrich_report_with_guidelines(scan_report, guidelines):
                 record.set_guideline(guidelines[record.check_id])
+
+    @staticmethod
+    def get_enriched_resources(report, repo_root):
```
Good catch. Removing.

ac-square commented on checkov/common/runners/runner_registry.py:
```diff
@@ -99,3 +114,74 @@ def enrich_report_with_guidelines(scan_report, guidelines):
+            out_parsing_errors=parsing_errors,
+        )
+
+        enriched_resources = defaultdict(dict)
```
If the resource isn't found in the repo-root, it won't get enriched. When `enriched_resources` is used in the `enrich_plan_report` function to lookup what the resource enrichments are, the defaultdict prevents a keyerror if that resource can't be found.

gruebel replied:
The reason to use `defaultdict` is correct, but your code won't raise any error, because of now using `enriched_resources.get(...)`
https://github.com/bridgecrewio/checkov/blob/9559952090aa517de3d49e5deb62a7134fc852b7/checkov/common/runners/runner_registry.py#L161-L162

ac-square replied:
Ack. Makes sense.

ac-square commented on checkov/common/runners/runner_registry.py:
```diff
@@ -99,3 +114,74 @@ def enrich_report_with_guidelines(scan_report, guidelines):
+                        entity_context = dict_utils.getInnerDict(
+                            definitions_context[full_file_path], entity_context_path
+                        )
+                        entity_lines_range = [
+                            entity_context.get("start_line"),
+                            entity_context.get("end_line"),
+                        ]
+                        entity_code_lines = entity_context.get("code_lines")
+                        skipped_checks = entity_context.get("skipped_checks")
```
Yes, that basically would mean the resource couldn't be found in the provided repo_root. That record would stay as-is, with the line numbers, code blocks, and filepath pointing to the provided plan file.

gruebel commented on checkov/common/runners/runner_registry.py:
```diff
@@ -99,3 +114,67 @@ def enrich_report_with_guidelines(scan_report, guidelines):
+    def handle_skipped_checks(report, enriched_resources):
+        records_to_skip = []
+        for record in report.failed_checks:
+            resource_skips = enriched_resources[record.resource]["skipped_checks"]
```
This can result in an exception, even using `defaultdict`. If the key `record.resource` doesn't exist, then `defaultdict` would return a dict, but then accessing `skipped_checks` key will raise a `KeyError`.

ac-square commented on checkov/common/runners/runner_registry.py:
```diff
@@ -99,3 +114,74 @@ def enrich_report_with_guidelines(scan_report, guidelines):
+                    record.check_result["result"] = CheckResult.SKIPPED
+                    record.check_result["suppress_comment"] = skip["suppress_comment"]
+                    report.add_record(record)
+        return report
```
Done. Moved them over to the `Report` class

ac-square commented:
Thanks @gruebel for the review. I've addressed all your comments and made the appropriate changes. I just have to write a few tests and then this should be good to go.

gruebel commented:
Looks good so far, now some tests and then it will be a great addition to `checkov` 💪

ac-square commented:
@schosterbarak @gruebel I believe this should be good to go now.

schosterbarak commented:
> @schosterbarak @gruebel I believe this should be good to go now.
One failing test. 
FAILED tests/common/runner_registry/test_runner_registry.py::TestRunnerRegistry::test_enrichment_of_plan_report
Let me know if you'd like me to dig into this one.

ac-square commented:
That's the test I just created. Was _just_ working locally. Let me debug.

nimrodkor reviewed (approved):
Thanks @ac-square !

Tests issue was due to a bug we had in master. It would seem the github checkout action merges the branch with the target branch before continuing, leading to different results then when running locally. Fix was deployed in #1253 , 🚀

christophetd commented:
@ac-square Just wanted to say a big thank you for the work on this PR. I've been waiting for this feature for several months and it's working *great*. Cheers!

larryboymi commented:
Using 1.7.5 works like a charm, thanks @ac-square and @christophetd for tracking this down. I was coming today to create this very issue (suppressing based on plan vs source HCL). Thanks!

larryboymi commented:
> > > > ignore external modules when presenting checkov results to end users because the modules are usually owned by another team (or are third
> > > 
> > > 
> > > @ac-square what do you show for plan scanning results if a resource is sourced from an external module? how do you point back to the code today? or is it just dropped from the results?
> > 
> > 
> > At Square we just drop them, but for this PR, they will be treated as normal results, just without the filepath/codeblock enrichment.
> 
> sounds good. thanks!

I would like if you could add a 2nd (or multiple) enrichment root for helping with external modules, as there is no current way that I can see to suppress checks that fail for them https://github.com/bridgecrewio/checkov/issues/560#issuecomment-881471743
